Add torch.Tensor._make_dtensor to accelerate DTensor.__new__ further

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #162220
* #162219
* #162218
* #161694
* #161693
* #161692
* #161634
* #161633
* #161596
* #161595
* #161591
* __->__ #161590
* #161586
* #161466

This seems to be a (very very roughly) ~8% improvement on DTensor benchmark very similar to the benchmark from #160580 (120ish usec -> 110ish usec)

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim

Differential Revision: [D81530105](https://our.internmc.facebook.com/intern/diff/D81530105)

swolchok committed:
Add torch.Tensor._make_dtensor to accelerate DTensor.__new__ further

This seems to be a (very very roughly) ~8% improvmeent on DTensor benchmark very similar to the benchmark from #160580 (120ish ms -> 110ish ms)

[ghstack-poisoned]

swolchok committed:
can't make memory-safety-critical checks debug-only on "Add torch.Tensor._make_dtensor to accelerate DTensor.__new__ further"


This seems to be a (very very roughly) ~8% improvmeent on DTensor benchmark very similar to the benchmark from #160580 (120ish ms -> 110ish ms)

[ghstack-poisoned]

swolchok commented on torch/distributed/tensor/_api.py:
```diff
@@ -270,23 +270,13 @@ def __new__(
-        )
-        r = torch.Tensor._make_wrapper_subclass(
+
+        r = torch.Tensor._make_dtensor(
```
@bdhirsh does this change also allow us to remove the torch._disable_dynamo annotation on this function? I'm still pretty unfamiliar with dynamo

bdhirsh replied:
although that reminds me - @anijain2305 I remember you telling me a long time ago that if we wanted a python function to be disabled with `recursive=True`, we have no way of doing that in dynamo and we're forced to do it via decorator on the python function itself.

We care a lot about DTensor overhead now, and this decorator is slow. If we're willing to hardcode some DTensor knowledge in dynamo, what do you think about effectively generated the `_disable_dynamo()` decorator around `DTensor.__new__` here in dynamo?

swolchok committed:
Update on "Add torch.Tensor._make_dtensor to accelerate DTensor.__new__ further"


This seems to be a (very very roughly) ~8% improvmeent on DTensor benchmark very similar to the benchmark from #160580 (120ish ms -> 110ish ms)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -771,6 +773,79 @@ static PyObject* THPVariable_make_wrapper_subclass(
   END_HANDLE_TH_ERRORS
 }
 
+// DTensor-specific variant of make_wrapper_subclass to minimize DTensor
```
Let's move this to a distributed folder?

swolchok replied:
which one? we have torch/csrc/distributed but none of the 3 subdirectories in there looks right to me at first glance

albanD commented on torch/csrc/Module.cpp:
```diff
@@ -2757,21 +2757,6 @@ Call this whenever a new thread is created in order to propagate values from
   ASSERT_TRUE(torch::disabled_torch_dispatch_impl() != nullptr);
 
-  // DTensor acceleration.
-  py_module.def(
```
Isn't this added lower in this stack?

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -771,6 +773,79 @@ static PyObject* THPVariable_make_wrapper_subclass(
 
+// DTensor-specific variant of make_wrapper_subclass to minimize DTensor
+// overhead.
+static PyObject* THPVariable_make_dtensor(
```
What's the difference with the original one? Is this whole thing to avoid one extra pybind round trip? That should be a few 100ns at best?

swolchok replied:
it is several pybind roundtrips -- each time the Python code messes with a DispatchKey or DispatchKeySet, that's another one.

swolchok replied:
creating a specialized _make_dtensor also allows us to avoid several (pybind!) property accesses on local_tensor that would otherwise be needed to call the generic _make_wrapper_subclass.

swolchok committed:
rebase on "Add torch.Tensor._make_dtensor to accelerate DTensor.__new__ further"


This seems to be a (very very roughly) ~8% improvmeent on DTensor benchmark very similar to the benchmark from #160580 (120ish ms -> 110ish ms)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

swolchok committed:
Update on "Add torch.Tensor._make_dtensor to accelerate DTensor.__new__ further"


This seems to be a (very very roughly) ~8% improvement on DTensor benchmark very similar to the benchmark from #160580 (120ish ms -> 110ish ms)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

ezyang commented:
@albanD did you want to make the final call on this one?

albanD reviewed (approved):
I guess we'll see how that goes going forward!

swolchok committed:
Update on "Add torch.Tensor._make_dtensor to accelerate DTensor.__new__ further"


This seems to be a (very very roughly) ~8% improvement on DTensor benchmark very similar to the benchmark from #160580 (120ish usec -> 110ish usec)

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

Differential Revision: [D81530105](https://our.internmc.facebook.com/intern/diff/D81530105)

[ghstack-poisoned]

swolchok commented:
@pytorchbot merge

swolchok commented:
@pytorchbot merge
